#  DATA CLEANING AND MANIPULATION

In [123]:
import pandas as pd
pd.set_option('display.max_columns', None)
import warnings
warnings.filterwarnings("ignore")

Importing the required libraries.

In [124]:
df = pd.read_csv("brooklyn_sales_map.csv", low_memory=False)

In [125]:
#checking the total rows and columns
df.shape

(390883, 111)

In [126]:
# checking the first 5 rows
df.head()

Unnamed: 0  borough           neighborhood building_class_category  \
0           1        3     DOWNTOWN-METROTECH   28  COMMERCIAL CONDOS   
1           2        3  DOWNTOWN-FULTON FERRY  29  COMMERCIAL GARAGES   
2           3        3       BROOKLYN HEIGHTS    21  OFFICE BUILDINGS   
3           4        3             MILL BASIN     22  STORE BUILDINGS   
4           5        3       BROOKLYN HEIGHTS         26 OTHER HOTELS   

  tax_class  block   lot  easement building_class              address  \
0         4    140  1001       NaN             R5       330 JAY STREET   
1         4     54     1       NaN             G7        85 JAY STREET   
2         4    204     1       NaN             O6  29 COLUMBIA HEIGHTS   
3         4   8470    55       NaN             K6        5120 AVENUE U   
4         4    230     1       NaN             H8      21 CLARK STREET   

  apartment_number  zip_code  residential_units  commercial_units  \
0            COURT     11201                  0                 1   
1              NaN     11201                  0                 0   
2              NaN     11201                  0                 0   
3              NaN     11234                  0               123   
4              NaN     11201                  0                 1   

   total_units  land_sqft  gross_sqft  year_built  tax_class_at_sale  \
0            1        0.0         0.0        2002                  4   
1            0   134988.0         0.0           0                  4   
2            0    32000.0    304650.0        1924                  4   
3          123   905000.0   2548000.0        1970                  4   
4            1    20267.0    356000.0        1927                  4   

  building_class_at_sale   sale_price   sale_date  year_of_sale Borough  \
0                     R5  499401179.0  2008-04-23          2008     NaN   
1                     G7  345000000.0  2016-12-20          2016      BK   
2                     O6  340000000.0  2016-08-03          2016      BK   
3                     K6  276947000.0  2012-11-28          2012      BK   
4                     H8  202500000.0  2017-10-31          2017      BK   

      CD  CT2010  CB2010  SchoolDist  Council  ZipCode FireComp  PolicePrct  \
0    NaN     NaN     NaN         NaN      NaN      NaN      NaN         NaN   
1  302.0   21.00  3017.0        13.0     33.0  11201.0     L118        84.0   
2  302.0    1.00  1003.0        13.0     33.0  11201.0     E205        84.0   
3  318.0  698.00  2005.0        22.0     46.0  11234.0     E323        63.0   
4  302.0    3.01  2000.0        13.0     33.0  11201.0     E205        84.0   

   HealthCent  HealthArea  SanitBoro  SanitDistr SanitSub  \
0         NaN         NaN        NaN         NaN      NaN   
1        36.0      1000.0        3.0         2.0       1B   
2        38.0      2300.0        3.0         2.0       1A   
3        35.0      8822.0        3.0        18.0       4E   
4        38.0      2300.0        3.0         2.0       1A   

               Address ZoneDist1 ZoneDist2 ZoneDist3 ZoneDist4 Overlay1  \
0                  NaN       NaN       NaN       NaN       NaN      NaN   
1        87 JAY STREET   M1-2/R8       NaN       NaN       NaN      NaN   
2  29 COLUMBIA HEIGHTS      M2-1       NaN       NaN       NaN      NaN   
3        5120 AVENUE U      M3-1       NaN       NaN       NaN      NaN   
4     79 WILLOW STREET        R6       NaN       NaN       NaN      NaN   

  Overlay2 SPDist1 SPDist2 SPDist3 LtdHeight SplitZone BldgClass  LandUse  \
0      NaN     NaN     NaN     NaN       NaN       NaN       NaN      NaN   
1      NaN    MX-2     NaN     NaN       NaN         N        G7     10.0   
2      NaN     NaN     NaN     NaN       NaN         N        O6      5.0   
3      NaN     NaN     NaN     NaN       NaN         N        K6      5.0   
4      NaN     NaN     NaN     NaN      LH-1         N        H8      8.0   

   Easements OwnerType              OwnerName   LotArea   BldgAr

The above dataframe has 111 columns but not all the columns are relevant to this project, so we decided to extract the columns needed.

In [127]:
subset_df = df[["neighborhood", "building_class_category", "tax_class_at_sale", "building_class_at_sale", 
          "address", "sale_price", "sale_date", "year_of_sale", "land_sqft", "gross_sqft", "OwnerName"]]

In [128]:
subset_df.dtypes

neighborhood                object
building_class_category     object
tax_class_at_sale            int64
building_class_at_sale      object
address                     object
sale_price                 float64
sale_date                   object
year_of_sale                 int64
land_sqft                  float64
gross_sqft                 float64
OwnerName                   object
dtype: object

Looking at the sale_date dtype, we  can see its not with the right dtype, needs to be changed to the right dtype.

In [129]:
subset_df["sale_date"] = pd.to_datetime(subset_df['sale_date'])

Sale_date is changed to its correct dtype and we need to extract the months seperately into a different column so we can use this column to show monthly trend afterwards.

In [130]:
subset_df['month'] = pd.DatetimeIndex(subset_df['sale_date']).month

In [131]:
subset_df.dtypes

neighborhood                       object
building_class_category            object
tax_class_at_sale                   int64
building_class_at_sale             object
address                            object
sale_price                        float64
sale_date                  datetime64[ns]
year_of_sale                        int64
land_sqft                         float64
gross_sqft                        float64
OwnerName                          object
month                               int64
dtype: object

In [132]:
subset_df.head()

neighborhood building_class_category  tax_class_at_sale  \
0     DOWNTOWN-METROTECH   28  COMMERCIAL CONDOS                  4   
1  DOWNTOWN-FULTON FERRY  29  COMMERCIAL GARAGES                  4   
2       BROOKLYN HEIGHTS    21  OFFICE BUILDINGS                  4   
3             MILL BASIN     22  STORE BUILDINGS                  4   
4       BROOKLYN HEIGHTS         26 OTHER HOTELS                  4   

  building_class_at_sale              address   sale_price  sale_date  \
0                     R5       330 JAY STREET  499401179.0 2008-04-23   
1                     G7        85 JAY STREET  345000000.0 2016-12-20   
2                     O6  29 COLUMBIA HEIGHTS  340000000.0 2016-08-03   
3                     K6        5120 AVENUE U  276947000.0 2012-11-28   
4                     H8      21 CLARK STREET  202500000.0 2017-10-31   

   year_of_sale  land_sqft  gross_sqft              OwnerName  month  
0          2008        0.0         0.0                    NaN      4  
1          2016   134988.0         0.0  85 JAY STREET BROOKLY     12  
2          2016    32000.0    304650.0  25-30 COLUMBIA HEIGHT      8  
3          2012   905000.0   2548000.0   BROOKLYN KINGS PLAZA     11  
4          2017    20267.0    356000.0  WATCHTOWER C/O REAL P     10

Looking at the month column, we see the records are in numeric from 1 to 12 which shows january to December, to be more explicit, its a good idea to change them to their respective month name.

In [133]:
subset_df.loc[subset_df["month"] == 1, "month"] = "January"

In [134]:
subset_df.loc[subset_df["month"] == 2, "month"] = "February"

In [135]:
subset_df.loc[subset_df["month"] == 3, "month"] = "March"

In [136]:
subset_df.loc[subset_df["month"] == 4, "month"] = "April"

In [137]:
subset_df.loc[subset_df["month"] == 5, "month"] = "May"

In [138]:
subset_df.loc[subset_df["month"] == 6, "month"] = "June"

In [139]:
subset_df.loc[subset_df["month"] == 7, "month"] = "July"

In [140]:
subset_df.loc[subset_df["month"] == 8, "month"] = "August"

In [141]:
subset_df.loc[subset_df["month"] == 9, "month"] = "September"

In [142]:
subset_df.loc[subset_df["month"] == 10, "month"] = "October"

In [143]:
subset_df.loc[subset_df["month"] == 11, "month"] = "November"

In [144]:
subset_df.loc[subset_df["month"] == 12, "month"] = "December"

In [145]:
# having a look at the change
subset_df.head()

neighborhood building_class_category  tax_class_at_sale  \
0     DOWNTOWN-METROTECH   28  COMMERCIAL CONDOS                  4   
1  DOWNTOWN-FULTON FERRY  29  COMMERCIAL GARAGES                  4   
2       BROOKLYN HEIGHTS    21  OFFICE BUILDINGS                  4   
3             MILL BASIN     22  STORE BUILDINGS                  4   
4       BROOKLYN HEIGHTS         26 OTHER HOTELS                  4   

  building_class_at_sale              address   sale_price  sale_date  \
0                     R5       330 JAY STREET  499401179.0 2008-04-23   
1                     G7        85 JAY STREET  345000000.0 2016-12-20   
2                     O6  29 COLUMBIA HEIGHTS  340000000.0 2016-08-03   
3                     K6        5120 AVENUE U  276947000.0 2012-11-28   
4                     H8      21 CLARK STREET  202500000.0 2017-10-31   

   year_of_sale  land_sqft  gross_sqft              OwnerName     month  
0          2008        0.0         0.0                    NaN     April  
1          2016   134988.0         0.0  85 JAY STREET BROOKLY  December  
2          2016    32000.0    304650.0  25-30 COLUMBIA HEIGHT    August  
3          2012   905000.0   2548000.0   BROOKLYN KINGS PLAZA  November  
4          2017    20267.0    356000.0  WATCHTOWER C/O REAL P   October

Now we have all columns needed , its time to check the null values in our dataframe.

In [146]:
subset_df.isnull().sum()

neighborhood                   0
building_class_category       83
tax_class_at_sale              0
building_class_at_sale         0
address                        1
sale_price                     0
sale_date                      0
year_of_sale                   0
land_sqft                      0
gross_sqft                     0
OwnerName                  87259
month                          0
dtype: int64

There are 3 columns with null values, however the OwnerName column which holds the name of individuals who bought this properties has the highest number of missing values, we assume their name are not recorded due to private reasons, so its reasonable to fill it with Anonymous. The building_class_category has 83 missing values, this will take some diggings to be able to figure out the class of the building missing and also for one record missing under the address. We will start with the building_class_category to see if we can get a pattern from the building_class_at_sale that can help us fill the building_class_category missing values.

In [147]:
#checking where building_class_category is null
subset_df.loc[subset_df["building_class_category"].isnull()]

neighborhood building_class_category  tax_class_at_sale  \
12659      WILLIAMSBURG-EAST                     NaN                  4   
24733     WILLIAMSBURG-NORTH                     NaN                  4   
29285   WILLIAMSBURG-CENTRAL                     NaN                  4   
55539              GRAVESEND                     NaN                  4   
152748            GREENPOINT                     NaN                  4   
...                      ...                     ...                ...   
214403    WILLIAMSBURG-NORTH                     NaN                  4   
214404    WILLIAMSBURG-NORTH                     NaN                  4   
214405    WILLIAMSBURG-NORTH                     NaN                  4   
214874    WILLIAMSBURG-SOUTH                     NaN                  4   
214875    WILLIAMSBURG-SOUTH                     NaN                  4   

       building_class_at_sale                 address  sale_price  sale_date  \
12659                      RK         1 POWERS STREET   1700000.0 2012-12-27   
24733                      RB   119 NORTH 11TH STREET   1175000.0 2012-12-27   
29285                      RK     505 FLUSHING AVENUE   1029014.0 2012-12-26   
55539                      RB            143 AVENUE O    735875.0 2012-12-31   
152748                     RB       231 NORMAN AVENUE    229106.0 2012-12-13   
...                       ...                     ...         ...        ...   
214403                     RW   22 NORTH 6TH   STREET         0.0 2012-12-10   
214404                     RG  101 NORTH 5TH   STREET         0.0 2012-12-26   
214405                     RB        287 GRAND STREET         0.0 2012-12-27   
214874                     RK        186 GRAND STREET         0.0 2012-12-27   
214875                     RG         24 DUNHAM PLACE         0.0 2012-11-30   

        year_of_sale  land_sqft  gross_sqft OwnerName     month  
12659           2012        0.0         0.0       NaN  December  
24733           2012        0.0         0.0       NaN  December  
29285           2012        0.0         0.0       NaN  December  
55539           2012        0.0         0.0       NaN  December  
152748          2012        0.0         0.0       NaN  December  
...              ...        ...         ...       ...       ...  
214403          2012        0.0         0.0       NaN  December  
214404          2012        0.0         0.0       NaN  December  
214405          2012        0.0         0.0       NaN  December  
214874          2012        0.0         0.0       NaN  December  
214875          2012        0.0         0.0       NaN  November  

[83 rows x 12 columns]

In [148]:
# looking at the distinct building_class_at_sale
subset_df.loc[subset_df["building_class_category"].isnull()].building_class_at_sale.value_counts()

RG    41
RP    14
RB    10
RS     9
RW     6
RK     3
Name: building_class_at_sale, dtype: int64

In [149]:
# taking the most frequent one into account
subset_df.loc[subset_df["building_class_at_sale"] == "RG", "building_class_category"]

1227      44  CONDO PARKING
1523      44  CONDO PARKING
2242      44  CONDO PARKING
2391       44 CONDO PARKING
4693      44  CONDO PARKING
                ...        
259113     44 CONDO PARKING
259115     44 CONDO PARKING
259169     44 CONDO PARKING
259170     44 CONDO PARKING
259172     44 CONDO PARKING
Name: building_class_category, Length: 2378, dtype: object

In [150]:
subset_df.loc[subset_df["building_class_category"].isnull(), "building_class_category"] = "44 CONDO PARKING"

In [151]:
subset_df.isnull().sum()

neighborhood                   0
building_class_category        0
tax_class_at_sale              0
building_class_at_sale         0
address                        1
sale_price                     0
sale_date                      0
year_of_sale                   0
land_sqft                      0
gross_sqft                     0
OwnerName                  87259
month                          0
dtype: int64

In [152]:
# checking where address is null
subset_df.loc[subset_df["address"].isnull()]

neighborhood          building_class_category  tax_class_at_sale  \
169571  SHEEPSHEAD BAY  10  COOPS - ELEVATOR APARTMENTS                  2   

       building_class_at_sale address  sale_price  sale_date  year_of_sale  \
169571                     D4     NaN    120000.0 2003-06-24          2003   

        land_sqft  gross_sqft              OwnerName month  
169571        0.0         0.0  ATLANTIC TOWERS APART  June

In [153]:
subset_df.loc[(subset_df["neighborhood"] == "SHEEPSHEAD BAY") & (subset_df["building_class_category"] == "10 COOPS - ELEVATOR APARTMENTS")].address.value_counts()

2285 BRAGG STREET, 4B       2
3060 OCEAN AVENUE, 4C       1
3090 VOORHIES AVENUE, 5F    1
3105 AVENUE V, 3H           1
2211 BRAGG STREET, 6B       1
                           ..
2455 HARING STREET, 6H      1
2800 COYLE STREET, 703      1
2951 OCEAN AVENUE, 1C       1
3060 OCEAN AVENUE, 2V       1
3080 VOORHIES AVENUE, 5H    1
Name: address, Length: 238, dtype: int64

In [154]:
# replacing null value in the address with "Not specified" 
subset_df.loc[subset_df["address"].isnull(), "address"] = "Not specified"

In [155]:
# replacing null value in the address with "Anonymous" 
subset_df.loc[subset_df["OwnerName"].isnull(), "OwnerName"] = "Anonymous"

In [156]:
subset_df.isnull().sum()

neighborhood               0
building_class_category    0
tax_class_at_sale          0
building_class_at_sale     0
address                    0
sale_price                 0
sale_date                  0
year_of_sale               0
land_sqft                  0
gross_sqft                 0
OwnerName                  0
month                      0
dtype: int64

In [157]:
# checking the distinct neigborhood
subset_df.neighborhood.value_counts().keys()

Index(['BEDFORD STUYVESANT', 'EAST NEW YORK', 'BOROUGH PARK', 'BUSHWICK',
       'CANARSIE', 'BAY RIDGE', 'CROWN HEIGHTS', 'FLATBUSH-EAST',
       'SHEEPSHEAD BAY', 'GRAVESEND', 'PARK SLOPE', 'SUNSET PARK',
       'BENSONHURST', 'FLATBUSH-CENTRAL', 'OCEAN PARKWAY-NORTH', 'OCEAN HILL',
       'WILLIAMSBURG-EAST', 'GREENPOINT', 'WILLIAMSBURG-NORTH',
       'BRIGHTON BEACH', 'MIDWOOD', 'CYPRESS HILLS', 'FLATBUSH-NORTH',
       'MADISON', 'MARINE PARK', 'OCEAN PARKWAY-SOUTH', 'CLINTON HILL',
       'BROOKLYN HEIGHTS', 'DYKER HEIGHTS', 'BATH BEACH', 'PROSPECT HEIGHTS',
       'WILLIAMSBURG-CENTRAL', 'WILLIAMSBURG-SOUTH', 'BROWNSVILLE',
       'OLD MILL BASIN', 'FLATLANDS', 'FORT GREENE', 'CONEY ISLAND',
       'PARK SLOPE SOUTH', 'CARROLL GARDENS', 'BERGEN BEACH',
       'DOWNTOWN-FULTON FERRY', 'BOERUM HILL', 'WINDSOR TERRACE', 'GOWANUS',
       'FLATBUSH-LEFFERTS GARDEN', 'KENSINGTON', 'DOWNTOWN-FULTON MALL',
       'DOWNTOWN-METROTECH', 'WYCKOFF HEIGHTS', 'GERRITSEN BEACH',
       'COBBL

The list above are the neighborhoods in brooklyn, we have 3 incorrect records (BROOKLYN-UNKNOWN, 3004, 3019), this certainly are not neighborhoods name in brooklyn. Time tio investigate the records so we can make changes to this incorrect neighborhoods.

In [158]:
# checking where neighborhood is 3004 and 3019
subset_df.loc[(subset_df["neighborhood"] == "3004") | (subset_df["neighborhood"] == "3019")]

neighborhood            building_class_category  tax_class_at_sale  \
269099         3019  08  RENTALS - ELEVATOR APARTMENTS                  2   
390881         3004             04  TAX CLASS 1 CONDOS                  1   
390882         3004             04  TAX CLASS 1 CONDOS                  1   

       building_class_at_sale                address  sale_price  sale_date  \
269099                     D1  649 WASHINGTON AVENUE         0.0 2007-12-12   
390881                     R3     14 BAY 47TH STREET    346788.0 2006-05-24   
390882                     R3       14 BAY 47 STREET    334620.0 2006-04-06   

        year_of_sale  land_sqft  gross_sqft  OwnerName     month  
269099          2007        0.0         0.0  Anonymous  December  
390881          2006        0.0         0.0  Anonymous       May  
390882          2006        0.0         0.0  Anonymous     April

In [159]:
subset_df.loc[subset_df["address"] == "14 BAY 47TH STREET"]

neighborhood building_class_category  tax_class_at_sale  \
99932     GRAVESEND   04 TAX CLASS 1 CONDOS                  1   
124789    GRAVESEND  04  TAX CLASS 1 CONDOS                  1   
390881         3004  04  TAX CLASS 1 CONDOS                  1   

       building_class_at_sale             address  sale_price  sale_date  \
99932                      R3  14 BAY 47TH STREET    480000.0 2017-01-26   
124789                     R3  14 BAY 47TH STREET    368000.0 2013-05-06   
390881                     R3  14 BAY 47TH STREET    346788.0 2006-05-24   

        year_of_sale  land_sqft  gross_sqft  OwnerName    month  
99932           2017        0.0         0.0  Anonymous  January  
124789          2013        0.0         0.0  Anonymous      May  
390881          2006        0.0         0.0  Anonymous      May

In [160]:
subset_df.loc[subset_df["neighborhood"] == "3004", "neighborhood"] = "GRAVESEND"

In [161]:
subset_df.loc[subset_df["address"] == "649 WASHINGTON AVENUE"]

neighborhood            building_class_category  tax_class_at_sale  \
56520   CROWN HEIGHTS     12  CONDOS - WALKUP APARTMENTS                  2   
269099           3019  08  RENTALS - ELEVATOR APARTMENTS                  2   

       building_class_at_sale                address  sale_price  sale_date  \
56520                      R2  649 WASHINGTON AVENUE    728558.0 2008-02-25   
269099                     D1  649 WASHINGTON AVENUE         0.0 2007-12-12   

        year_of_sale  land_sqft  gross_sqft  OwnerName     month  
56520           2008        0.0         0.0  Anonymous  February  
269099          2007        0.0         0.0  Anonymous  December

In [162]:
subset_df.loc[subset_df["neighborhood"] == "3019", "neighborhood"] = "CROWN HEIGHTS"

In [163]:
subset_df.loc[subset_df["neighborhood"] == 'BROOKLYN-UNKNOWN', "neighborhood"] = "BROOKLYN HEIGHTS"

In [164]:
subset_df.neighborhood.value_counts().keys()

Index(['BEDFORD STUYVESANT', 'EAST NEW YORK', 'BOROUGH PARK', 'BUSHWICK',
       'CANARSIE', 'BAY RIDGE', 'CROWN HEIGHTS', 'FLATBUSH-EAST',
       'SHEEPSHEAD BAY', 'GRAVESEND', 'PARK SLOPE', 'SUNSET PARK',
       'BENSONHURST', 'FLATBUSH-CENTRAL', 'OCEAN PARKWAY-NORTH', 'OCEAN HILL',
       'WILLIAMSBURG-EAST', 'GREENPOINT', 'WILLIAMSBURG-NORTH',
       'BRIGHTON BEACH', 'MIDWOOD', 'CYPRESS HILLS', 'FLATBUSH-NORTH',
       'MADISON', 'MARINE PARK', 'OCEAN PARKWAY-SOUTH', 'CLINTON HILL',
       'BROOKLYN HEIGHTS', 'DYKER HEIGHTS', 'BATH BEACH', 'PROSPECT HEIGHTS',
       'WILLIAMSBURG-CENTRAL', 'WILLIAMSBURG-SOUTH', 'BROWNSVILLE',
       'OLD MILL BASIN', 'FLATLANDS', 'FORT GREENE', 'CONEY ISLAND',
       'PARK SLOPE SOUTH', 'CARROLL GARDENS', 'BERGEN BEACH',
       'DOWNTOWN-FULTON FERRY', 'BOERUM HILL', 'WINDSOR TERRACE', 'GOWANUS',
       'FLATBUSH-LEFFERTS GARDEN', 'KENSINGTON', 'DOWNTOWN-FULTON MALL',
       'DOWNTOWN-METROTECH', 'WYCKOFF HEIGHTS', 'GERRITSEN BEACH',
       'COBBL

All incorrect neihborhood are corrected, next is to classify the neighborhood into theor geographical location which will take its own colum. With the help of wikipidia we can get their respective georgraphical region. https://en.wikipedia.org/wiki/List_of_Brooklyn_neighborhoods

In [165]:
# creating the geographical column and populating it with null values
subset_df = subset_df.assign(geographical_region = pd.NA)

In [166]:
subset_df.head()

neighborhood building_class_category  tax_class_at_sale  \
0     DOWNTOWN-METROTECH   28  COMMERCIAL CONDOS                  4   
1  DOWNTOWN-FULTON FERRY  29  COMMERCIAL GARAGES                  4   
2       BROOKLYN HEIGHTS    21  OFFICE BUILDINGS                  4   
3             MILL BASIN     22  STORE BUILDINGS                  4   
4       BROOKLYN HEIGHTS         26 OTHER HOTELS                  4   

  building_class_at_sale              address   sale_price  sale_date  \
0                     R5       330 JAY STREET  499401179.0 2008-04-23   
1                     G7        85 JAY STREET  345000000.0 2016-12-20   
2                     O6  29 COLUMBIA HEIGHTS  340000000.0 2016-08-03   
3                     K6        5120 AVENUE U  276947000.0 2012-11-28   
4                     H8      21 CLARK STREET  202500000.0 2017-10-31   

   year_of_sale  land_sqft  gross_sqft              OwnerName     month  \
0          2008        0.0         0.0              Anonymous     April   
1          2016   134988.0         0.0  85 JAY STREET BROOKLY  December   
2          2016    32000.0    304650.0  25-30 COLUMBIA HEIGHT    August   
3          2012   905000.0   2548000.0   BROOKLYN KINGS PLAZA  November   
4          2017    20267.0    356000.0  WATCHTOWER C/O REAL P   October   

  geographical_region  
0                <NA>  
1                <NA>  
2                <NA>  
3                <NA>  
4                <NA>

In [167]:
subset_df.neighborhood.value_counts().keys()

Index(['BEDFORD STUYVESANT', 'EAST NEW YORK', 'BOROUGH PARK', 'BUSHWICK',
       'CANARSIE', 'BAY RIDGE', 'CROWN HEIGHTS', 'FLATBUSH-EAST',
       'SHEEPSHEAD BAY', 'GRAVESEND', 'PARK SLOPE', 'SUNSET PARK',
       'BENSONHURST', 'FLATBUSH-CENTRAL', 'OCEAN PARKWAY-NORTH', 'OCEAN HILL',
       'WILLIAMSBURG-EAST', 'GREENPOINT', 'WILLIAMSBURG-NORTH',
       'BRIGHTON BEACH', 'MIDWOOD', 'CYPRESS HILLS', 'FLATBUSH-NORTH',
       'MADISON', 'MARINE PARK', 'OCEAN PARKWAY-SOUTH', 'CLINTON HILL',
       'BROOKLYN HEIGHTS', 'DYKER HEIGHTS', 'BATH BEACH', 'PROSPECT HEIGHTS',
       'WILLIAMSBURG-CENTRAL', 'WILLIAMSBURG-SOUTH', 'BROWNSVILLE',
       'OLD MILL BASIN', 'FLATLANDS', 'FORT GREENE', 'CONEY ISLAND',
       'PARK SLOPE SOUTH', 'CARROLL GARDENS', 'BERGEN BEACH',
       'DOWNTOWN-FULTON FERRY', 'BOERUM HILL', 'WINDSOR TERRACE', 'GOWANUS',
       'FLATBUSH-LEFFERTS GARDEN', 'KENSINGTON', 'DOWNTOWN-FULTON MALL',
       'DOWNTOWN-METROTECH', 'WYCKOFF HEIGHTS', 'GERRITSEN BEACH',
       'COBBL

The neighborhoods above will be mapped to their respective Geographical region with the long list of code below.

In [168]:
subset_df.loc[subset_df["neighborhood"] == 'BEDFORD STUYVESANT', "geographical_region"] = "Northern Brooklyn"

In [169]:
subset_df.loc[subset_df["neighborhood"] == 'EAST NEW YORK', "geographical_region"] = "Eastern Brooklyn"

In [170]:
subset_df.loc[subset_df["neighborhood"] == 'BOROUGH PARK', "geographical_region"] = "Southwestern Brooklyn"

In [171]:
subset_df.loc[subset_df["neighborhood"] == 'BUSHWICK', "geographical_region"] = "Northern Brooklyn"

In [172]:
subset_df.loc[subset_df["neighborhood"] == 'CANARSIE', "geographical_region"] = "Eastern Brooklyn"

In [173]:
subset_df.loc[subset_df["neighborhood"] == 'BAY RIDGE', "geographical_region"] = "Southwestern Brooklyn"

In [174]:
subset_df.loc[subset_df["neighborhood"] == 'CROWN HEIGHTS', "geographical_region"] = "Central Brooklyn"

In [175]:
subset_df.loc[subset_df["neighborhood"] == 'FLATBUSH-EAST', "geographical_region"] = "Central Brooklyn"

In [176]:
subset_df.loc[subset_df["neighborhood"] == 'SHEEPSHEAD BAY', "geographical_region"] = "Southern Brooklyn"

In [177]:
subset_df.loc[subset_df["neighborhood"] == 'GRAVESEND', "geographical_region"] = "Southern Brooklyn"

In [178]:
subset_df.loc[subset_df["neighborhood"] == 'PARK SLOPE', "geographical_region"] = "Northwestern Brooklyn"

In [179]:
subset_df.loc[subset_df["neighborhood"] == 'SUNSET PARK', "geographical_region"] = "Southwestern Brooklyn"

In [180]:
subset_df.loc[subset_df["neighborhood"] == 'BENSONHURST', "geographical_region"] = "Southwestern Brooklyn"

In [181]:
subset_df.loc[subset_df["neighborhood"] == 'FLATBUSH-CENTRAL', "geographical_region"] = "Central Brooklyn"

In [182]:
subset_df.loc[subset_df["neighborhood"] == 'OCEAN PARKWAY-NORTH', "geographical_region"] = "Central Brooklyn"

In [183]:
subset_df.loc[subset_df["neighborhood"] == 'OCEAN HILL', "geographical_region"] = "Northern Brooklyn"

In [184]:
subset_df.loc[subset_df["neighborhood"] == 'WILLIAMSBURG-EAST', "geographical_region"] = "Northern Brooklyn"

In [185]:
subset_df.loc[subset_df["neighborhood"] == 'GREENPOINT', "geographical_region"] = "Northern Brooklyn"

In [186]:
subset_df.loc[subset_df["neighborhood"] == 'WILLIAMSBURG-NORTH', "geographical_region"] = "Northern Brooklyn"

In [187]:
subset_df.loc[subset_df["neighborhood"] == 'BRIGHTON BEACH', "geographical_region"] = "Southern Brooklyn"

In [188]:
subset_df.loc[subset_df["neighborhood"] == 'MIDWOOD', "geographical_region"] = "Southern Brooklyn"

In [189]:
subset_df.loc[subset_df["neighborhood"] == 'CYPRESS HILLS', "geographical_region"] = "Eastern Brooklyn"

In [190]:
subset_df.loc[subset_df["neighborhood"] == 'FLATBUSH-NORTH', "geographical_region"] = "Central Brooklyn"

In [191]:
subset_df.loc[subset_df["neighborhood"] == 'MADISON', "geographical_region"] = "Southern Brooklyn"

In [192]:
subset_df.loc[subset_df["neighborhood"] == 'MARINE PARK', "geographical_region"] = "Southern Brooklyn"

In [193]:
subset_df.loc[subset_df["neighborhood"] == 'OCEAN PARKWAY-SOUTH', "geographical_region"] = "Central Brooklyn"

In [194]:
subset_df.loc[subset_df["neighborhood"] == 'CLINTON HILL', "geographical_region"] = "Northwestern Brooklyn"

In [195]:
subset_df.loc[subset_df["neighborhood"] == 'BROOKLYN HEIGHTS', "geographical_region"] = "Northwestern Brooklyn"

In [196]:
subset_df.loc[subset_df["neighborhood"] == 'DYKER HEIGHTS', "geographical_region"] = "Southwestern Brooklyn"

In [197]:
subset_df.loc[subset_df["neighborhood"] == 'BATH BEACH', "geographical_region"] = "Southwestern Brooklyn"

In [198]:
subset_df.loc[subset_df["neighborhood"] == 'PROSPECT HEIGHTS', "geographical_region"] = "Northwestern Brooklyn"

In [199]:
subset_df.loc[subset_df["neighborhood"] == 'WILLIAMSBURG-CENTRAL', "geographical_region"] = "Northern Brooklyn"

In [200]:
subset_df.loc[subset_df["neighborhood"] == 'WILLIAMSBURG-SOUTH', "geographical_region"] = "Northern Brooklyn"

In [201]:
subset_df.loc[subset_df["neighborhood"] == 'BROWNSVILLE', "geographical_region"] = "Eastern Brooklyn"

In [202]:
subset_df.loc[subset_df["neighborhood"] == 'OLD MILL BASIN', "geographical_region"] = "Southern Brooklyn"

In [203]:
subset_df.loc[subset_df["neighborhood"] == 'FLATLANDS', "geographical_region"] = "Southern Brooklyn"

In [204]:
subset_df.loc[subset_df["neighborhood"] == 'FORT GREENE', "geographical_region"] = "Northwestern Brooklyn"

In [205]:
subset_df.loc[subset_df["neighborhood"] == 'CONEY ISLAND', "geographical_region"] = "Southern Brooklyn"

In [206]:
subset_df.loc[subset_df["neighborhood"] == 'PARK SLOPE SOUTH', "geographical_region"] = "Northwestern Brooklyn"

In [207]:
subset_df.loc[subset_df["neighborhood"] == 'CARROLL GARDENS', "geographical_region"] = "Northwestern Brooklyn"

In [208]:
subset_df.loc[subset_df["neighborhood"] == 'BERGEN BEACH', "geographical_region"] = "Southern Brooklyn"

In [209]:
subset_df.loc[subset_df["neighborhood"] == 'DOWNTOWN-FULTON FERRY', "geographical_region"] = "Northwestern Brooklyn"

In [210]:
subset_df.loc[subset_df["neighborhood"] == 'BOERUM HILL', "geographical_region"] = "Northwestern Brooklyn"

In [211]:
subset_df.loc[subset_df["neighborhood"] == 'WINDSOR TERRACE', "geographical_region"] = "Central Brooklyn"

In [212]:
subset_df.loc[subset_df["neighborhood"] == 'GOWANUS', "geographical_region"] = "Northwestern Brooklyn"

In [213]:
subset_df.loc[subset_df["neighborhood"] == 'FLATBUSH-LEFFERTS GARDEN', "geographical_region"] = "Central Brooklyn"

In [214]:
subset_df.loc[subset_df["neighborhood"] == 'KENSINGTON', "geographical_region"] = "Central Brooklyn"

In [215]:
subset_df.loc[subset_df["neighborhood"] == 'DOWNTOWN-FULTON MALL', "geographical_region"] = "Northwestern Brooklyn"

In [216]:
subset_df.loc[subset_df["neighborhood"] == 'DOWNTOWN-METROTECH', "geographical_region"] = "Northwestern Brooklyn"

In [217]:
subset_df.loc[subset_df["neighborhood"] == 'WYCKOFF HEIGHTS', "geographical_region"] = "Northern Brooklyn"

In [218]:
subset_df.loc[subset_df["neighborhood"] == 'GERRITSEN BEACH', "geographical_region"] = "Southern Brooklyn"

In [219]:
subset_df.loc[subset_df["neighborhood"] == 'COBBLE HILL', "geographical_region"] = "Northwestern Brooklyn"

In [220]:
subset_df.loc[subset_df["neighborhood"] == 'MILL BASIN', "geographical_region"] = "Southern Brooklyn"

In [221]:
subset_df.loc[subset_df["neighborhood"] == 'COBBLE HILL-WEST', "geographical_region"] = "Northwestern Brooklyn"

In [222]:
subset_df.loc[subset_df["neighborhood"] == 'RED HOOK', "geographical_region"] = "Northwestern Brooklyn"

In [223]:
subset_df.loc[subset_df["neighborhood"] == 'MANHATTAN BEACH', "geographical_region"] = "Southern Brooklyn"

In [224]:
subset_df.loc[subset_df["neighborhood"] == 'BUSH TERMINAL', "geographical_region"] = "Southwestern Brooklyn"

In [225]:
subset_df.loc[subset_df["neighborhood"] == 'SEAGATE', "geographical_region"] = "Southern Brooklyn"

In [226]:
subset_df.loc[subset_df["neighborhood"] == 'SPRING CREEK', "geographical_region"] = "Eastern Brooklyn"

In [227]:
subset_df.loc[subset_df["neighborhood"] == 'NAVY YARD', "geographical_region"] = "Northwestern Brooklyn"

In [228]:
# having a look at the changes
subset_df.head()

neighborhood building_class_category  tax_class_at_sale  \
0     DOWNTOWN-METROTECH   28  COMMERCIAL CONDOS                  4   
1  DOWNTOWN-FULTON FERRY  29  COMMERCIAL GARAGES                  4   
2       BROOKLYN HEIGHTS    21  OFFICE BUILDINGS                  4   
3             MILL BASIN     22  STORE BUILDINGS                  4   
4       BROOKLYN HEIGHTS         26 OTHER HOTELS                  4   

  building_class_at_sale              address   sale_price  sale_date  \
0                     R5       330 JAY STREET  499401179.0 2008-04-23   
1                     G7        85 JAY STREET  345000000.0 2016-12-20   
2                     O6  29 COLUMBIA HEIGHTS  340000000.0 2016-08-03   
3                     K6        5120 AVENUE U  276947000.0 2012-11-28   
4                     H8      21 CLARK STREET  202500000.0 2017-10-31   

   year_of_sale  land_sqft  gross_sqft              OwnerName     month  \
0          2008        0.0         0.0              Anonymous     April   
1          2016   134988.0         0.0  85 JAY STREET BROOKLY  December   
2          2016    32000.0    304650.0  25-30 COLUMBIA HEIGHT    August   
3          2012   905000.0   2548000.0   BROOKLYN KINGS PLAZA  November   
4          2017    20267.0    356000.0  WATCHTOWER C/O REAL P   October   

     geographical_region  
0  Northwestern Brooklyn  
1  Northwestern Brooklyn  
2  Northwestern Brooklyn  
3      Southern Brooklyn  
4  Northwestern Brooklyn

In [229]:
subset_df.isnull().sum()

neighborhood               0
building_class_category    0
tax_class_at_sale          0
building_class_at_sale     0
address                    0
sale_price                 0
sale_date                  0
year_of_sale               0
land_sqft                  0
gross_sqft                 0
OwnerName                  0
month                      0
geographical_region        7
dtype: int64

After checking the null values above, we can see the geographical region has 7 missing values which weintentionally made it this way because of unclarity with one of the neighborhood(JAMAICA BAY). Politically, Jamaica bay seems not to be a neigborhood in Brooklyn, it is primarily divided between the boroughs of Brooklyn and Queens in New York City, with a small part touching Nassau County, so on this basis we decided not to fill it with geographical region we are not certain of, hence we decide we will drop it.

In [230]:
# checking where neighborhood is jamaica bay so we have a look of how many rows
subset_df.loc[subset_df["geographical_region"].isnull()]

neighborhood               building_class_category  tax_class_at_sale  \
740     JAMAICA BAY                   22  STORE BUILDINGS                  4   
3137    JAMAICA BAY            31  COMMERCIAL VACANT LAND                  4   
182078  JAMAICA BAY                 29 COMMERCIAL GARAGES                  4   
183021  JAMAICA BAY                   22  STORE BUILDINGS                  4   
183022  JAMAICA BAY            31  COMMERCIAL VACANT LAND                  4   
220632  JAMAICA BAY                   22  STORE BUILDINGS                  4   
220633  JAMAICA BAY  40  SELECTED GOVERNMENTAL FACILITIES                  4   

       building_class_at_sale               address  sale_price  sale_date  \
740                        K1  2875 FLATBUSH AVENUE  13000000.0 2013-02-21   
3137                       V9       FLATBUSH AVENUE   4200000.0 2013-03-14   
182078                     G8  2859 FLATBUSH AVENUE        10.0 2017-08-03   
183021                     K1  2875 FLATBUSH AVENUE         1.0 2013-02-21   
183022                     V9       FLATBUSH AVENUE         1.0 2013-03-14   
220632                     K1  2875 FLATBUSH AVENUE         0.0 2013-03-14   
220633                     Y7       FLATBUSH AVENUE         0.0 2013-03-14   

        year_of_sale  land_sqft  gross_sqft              OwnerName     month  \
740             2013   130120.0     45969.0  NEW YORK CITY LAND DE  February   
3137            2013   110303.0         0.0  BICAL DEVELOPMENT, IN     March   
182078          2017   110303.0    147187.0  BICAL DEVELOPMENT, IN    August   
183021          2013   130120.0     45969.0  NEW YORK CITY LAND DE  February   
183022          2013   110303.0         0.0  BICAL DEVELOPMENT, IN     March   
220632          2013   130120.0     45969.0  NEW YORK CITY LAND DE     March   
220633          2013   267417.0         0.0  TOYS R US-DELAWARE IN     March   

       geographical_region  
740                   <NA>  
3137                  <NA>  
182078                <NA>  
183021                <NA>  
183022                <NA>  
220632                <NA>  
220633                <NA>

In [231]:
#its not a lot of row, so we drop this rows
subset_df = subset_df.dropna()

In [232]:
subset_df.isnull().sum()

neighborhood               0
building_class_category    0
tax_class_at_sale          0
building_class_at_sale     0
address                    0
sale_price                 0
sale_date                  0
year_of_sale               0
land_sqft                  0
gross_sqft                 0
OwnerName                  0
month                      0
geographical_region        0
dtype: int64

Good that we no longer have null values, however in our dataframe under the sale_price columns, there are properties sale price that are less than $1000 and many more, so we decided to set a pricing threshold of 100000 above so we can use this in our analyis.

In [233]:
subset_df = subset_df.loc[subset_df["sale_price"] > 100000]

In [234]:
subset_df.shape

(233805, 13)

In [235]:
# saving the final dataframe to a csv so it can be use for Exploratory analysis.
subset_df.to_csv("subset_properties.csv")